WIP: Add behaviour for dealing with duplicate sites in Gutenboarding

andrewserong opened the pull request:
***Note:*** This PR originally solved the issue by adding a modal. The PR has since been simplified to deal with redirect behaviour implicitly.

---

This PR is an attempt at solving the duplicate sites in signup issue for Gutenboarding by storing the domain of the last created site in state. When Gutenboarding loads, if the last created site is recent (within 60 minutes) and the user is not on the first step of the flow (e.g. they landed there by hitting the back button), then we redirect to the customer home. If instead they navigated to the beginning of the flow (e.g. they clicked "Add a new site"), we clear the state and begin a fresh flow.

After the 60 minutes have elapsed, if a user navigates to Gutenboarding and a last created site exists in state, the Gutenboarding state is reset, and the user is redirected to the beginning of the signup flow, irrespective of which step they were on.

#### Changes proposed in this Pull Request

* Add logic to the Gutenboard component, that redirects the user to the Customer Home given the following conditions:
  * Haven't checked for a potential duplicate site yet
  * The user is logged in (there is a `currentUser`)
  * The last site was created recently (within the past 60 minutes)
  * The site exists
  * The user landed on a part of Gutenboarding that is _not_ the first step (e.g. they reached Gutenboarding via hitting the Back button)

* If the user begins at the first step of Gutenboarding (e.g. from clicking "Add a site") or the last site wasn't created recently, then we clear the onboarding store (and last created site state) and start from the beginning.

* Add a selector and resolver to the `site` data store `getSite` to pass in a domain/slug string and check against the `/sites` endpoint whether or not the site exists. This stores sites in an object, keyed by the site's slug.

#### Animated gif

Going through the following steps:

* From a freshly created site, the user clicks back
* They land on the last step of the Gutenboarding flow, and are then redirected to the Customer Home

![duplicate-site-redirect](https://user-images.githubusercontent.com/14988353/76391027-a74f0380-63c2-11ea-8555-f32e52059f7e.gif)

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to `/gutenboarding` and create a new site.
* Click back and you should be taken to the Customer Home for that site.
* Also create a new site, and then manually navigate to `/gutenboarding`, simulating what it might look like for a user who creates a site and then goes to "Add a new site" from the Calypso admin — it should begin a new site creation flow like normal.

Fixes #39268

andrewserong commented:
Added `lastCreatedSite` to persisted state so this is now technically working. Also added a check to see whether there is a current user before showing the modal. My plan for tomorrow is to do a bit more tweaking, and add in a check to see if the site _really_ exists before showing the modal, and then update the PR with screenshots to get some additional eyes and opinions on this approach.

andrewserong commented:
Apologies for the pings (if anyone's been pinged) ☝️was due to a bad rebase. Should be fixed now.

andrewserong commented:
This PR is now ready for a preliminary review / feedback. I'm keen to hear any ideas or thoughts on the approach — I thought it'd be a good idea to get something that we can play with and get a feel for before embarking on finessing the design. Also happy to go with a different approach to the modal — the main point is to deal with checking whether or not there is an existing site when Gutenboarding loads, and if so, do something that will make our users' lives easier!

ramonjd commented:
Thanks for taking this one on @andrewserong 🙇 

Works as described in the test instructions!

I've only done some prelim testing, but I found that clicking Create a new site in the modal, then clicking back to `/create-site` from there leaves me stranded and staring at those pulsating grey biscuits of doom. I'm not sure if another site is created or not.

There is probably a very good reason, but I'm wondering why `/create-site` is a separate route anyway? 

Is it meant to be the new [Processing Screen](https://github.com/Automattic/wp-calypso/blob/master/client/signup/processing-screen/index.jsx)?

Interested to hear what @Automattic/dotcom-create-design thinks overall.

andrewserong commented on client/landing/gutenboarding/stores/onboard/selectors.ts:
```diff
@@ -4,3 +4,19 @@
+ * @param {State} state		Global state tree
+ * @returns {boolean}		true if activity is in progress
+ */
+export const isLastCreatedSiteCurrent = ( state: State ): boolean => {
```
Good call! Updated.

andrewserong commented on client/landing/gutenboarding/stores/onboard/reducer.ts:
```diff
@@ -70,8 +70,19 @@ const pageLayouts: Reducer< string[], OnboardAction > = ( state = [], action ) =
+	action
+) => {
+	if ( action.type === 'RESET_ONBOARD_STORE' ) {
+		return action.lastCreatedSite || undefined;
```
Oh, yes, good point. Removed.

andrewserong commented on client/landing/gutenboarding/components/duplicate-site-modal/index.tsx:
```diff
@@ -0,0 +1,71 @@
+					{ NO__( 'Create new site' ) }
+				</Button>
+
+				<a className="duplicate-site-modal__cta-link" href={ `/home/${ lastCreatedSite?.domain }` }>
```
That looks much clearer, I've updated it.

andrewserong commented on packages/data-stores/src/site/resolvers.ts:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @param domain {string}	The domain to search for
+ */
+export async function getExistingSite( domain: string ) {
```
Done.

andrewserong commented:
Thanks for reviewing @p-jackson and the idea for the improved text, you're right, I think it needs something much friendlier and clearer like in your mockup! I've done a quick pass on the easy to fix immediate issues while I pop my head up from the data migration work. I'll continue on with the rest and updating the text tomorrow. Cheers!

andrewserong commented on client/landing/gutenboarding/components/duplicate-site-modal/index.tsx:
```diff
@@ -0,0 +1,71 @@
+	useEffect( () => {
+		if ( existingSite ) {
+			setShowDuplicateSiteModal( true );
+		}
```
Thanks! Yes, you're right it doesn't need to be here at all. I think I was torn about where to put the logic and then got stuck half way in between two approaches. I'll have a go at moving the check to `gutenboarding.tsx` and pass down the existing site in props so that the modal doesn't have to be aware of state (this conditional was originally there to prevent the modal from rendering empty strings if the details of the last created site weren't resolved yet).

ramonjd commented:
> I had another thought, which was that the store can only hold a single "existing site" at a time. Does that mean if two components call getExistingSite() with different arguments it will keep making a network request for each render?

Without thinking too deeply about it, this makes good sense to me. I guess it wouldn't be such a long shot to create middleware that detects that a site has been created in Gutenboarding and/or all Gutenboarding routes.

I'm not really sold on the create site step however.  I can't articulate why exactly. 

The general logic of the modal and the [last created check](https://github.com/Automattic/wp-calypso/pull/39794/files#diff-655ccf8d97a75836ef55d31cf1bbdf12R17), I think, makes sense.

Only that it seems like we'd have to continually play defence to avoid people navigating to it when they're not supposed to and becoming "stuck", as happens with the current processing screen (quite a lot).

It might be just that we're missing some pieces in the create step route, which I see doesn't house much business logic in itself.

olaolusoga commented:
Thanks for putting this together @p-jackson , and for the ping @ramonjd .

Question: Why do we need the modal?

In the event we created a site for the customer, if they click the back button before the 10 minutes elapses, we should automatically move them back to the created site. The modal seems unnecessary. We shouldn't allow folks to go back into the onboarding flow. If they'd like to create a new site, then they should select "Add new site" within the left side nav.

andrewserong commented:
Thanks for taking a look @olaolusoga!

> Question: Why do we need the modal?

That's a good question. It's probably not necessary, but I think it's ultimately a choice between introducing implicit versus explicit behaviour here. My personal preference (and I might be alone in this 😄) is for explicitly telling the user what we're about to do. If a user clicks the back button from the editor to get back to the site creation flow, but are then redirected back to the editor, is that confusing? And if someone who has just created a site goes to the sidebar wanting to make a change to their existing site and clicks "Add new site" (e.g. when they want to add a domain to the site they created), is it confusing for them to be starting at the beginning of the site creation flow again?

In either case it seems like the user might be trying or struggling to do a particular task, so I was wondering if a modal might help orient the user in these cases.

However, that's certainly not the MVP of dealing with duplicate sites. It's probably easier to implement the implicit way of handling it though, something like:

* When Gutenboarding loads, if we're on any step other than the first step / root of the flow, then check for a recently created site and if so, redirect to the editor (modelling the behaviour if the user clicked the back button).
* When Gutenboarding loads, if we're on the first step / root of the flow, and there's a recently created site in the state, then clear the state and start fresh.

I'm happy to re-work the PR, removing the modal but preserving the logic for storing the recently created site and implementing it in the implicit way. If it ends up being confusing for users further down the track, it won't be hard to add a modal back in.

andrewserong commented:
> For just the gutenboarding case what we've got works, but for a generic site store maybe the selector would be called getSite() and it'd keep all the returned sites in the store key'd by site slug?

@p-jackson that's a great idea, I'll have a play with this

> I'm not really sold on the create site step however. I can't articulate why exactly.

@ramonjd me too — a bit like the work Phil did in turning the account creation into a modal, the loading screen feels like it would better belong as a state rather than a route. Shall we look at changing that separately?

ramonjd commented:
> Shall we look at changing that separately?

Sure! Love iterations 🤗

andrewserong commented:
I've updated this WIP with the following:

* Changed the `getExistingSite` selector to `getSite` and updated the corresponding action and reducer to use an object of sites keyed by the site's slug. This feels like it's working much nicer now @p-jackson
* Updated the behaviour so that there is no modal, and instead we deal with redirect behaviour in the Gutenboard component implicitly. E.g. if a user lands on any step other than the first step in the flow, and the last created site is recent, we redirect to the Customer Home, otherwise we start a fresh onboarding flow.

Does this logic make sense? Does it feel okay if you test it out? I've updated the PR description with an animated GIF and to reflect removing the modal.

Happy for any other feedback — if it looks like an okay direction for this, I'll work on adding some tests for the site store and flip this PR over to `Ready for review` tomorrow.

p-jackson commented:
> We shouldn't allow folks to go back into the onboarding flow after they've been through it. If they'd like to create a new site, then they should select "Add new site" within the left side nav.

I'm worried that not allowing users to use the back button will cause frustration. I know when I encounter this behaviour on the web I find it pretty hostile, and I often ctrl+click the back button to try and skip past the page that's "misbehaving" (from my point of view).

The user may want to go back because they feel they've misclicked, or they're second guessing the choice they made, or they just want to compare the preview they saw earlier with what they're seeing in the editor.

Now I don't think we need to support all of those examples, but I don't think the user is "wrong" for trying something that seemed reasonable to them. Which is why I like the modal (or something that achieves the same thing), because it gives us an opportunity to tell the user that to make the changes they're trying to make they can't use the UI they just learnt and will need to press forward.

IMO, ideally they'd be able to back out all the way to intent gathering and change their site title. Which might have been possible if we didn't require an account before creating the site, but alas :'(

@olaolusoga

olaolusoga commented:
> IMO, ideally they'd be able to back out all the way to intent gathering and change their site title. Which might have been possible if we didn't require an account before creating the site, but alas :'(

That's changing with the new onboarding work (gutenboarding). I hear you on the frustration.

The modal still seems unnecessary. If a person goes back, and we want to stop site duplication, then is it possible to not create a new site?

Scenario: a person goes through onnboarding, lands in the dashboard, selects back, we allow them to go back, since we know they just created a site whatever adjustments they make is then reflected in the existing site, not in an entirely new site.

davipontesblog commented:
> a person goes through onnboarding, lands in the dashboard, selects back, we allow them to go back, since we know they just created a site whatever adjustments they make is then reflected in the existing site, not in an entirely new site.

>  If they'd like to create a new site, then they should select "Add new site" within the left side nav.

I think we need to think about why they are hitting the back button here. I think there are three reasons why they do that:
- I don't know what this screen is, let me go back to where I was and see what I did to land here.
- I don't' like what I see, and want to start over.
- I want to change something I chose in the site creation step, maybe the title, maybe the design, etc.

So to answer those questions we can simplify by offering two options:

- Start a new site entirely, and scrap the existing one
- Go back to "Site builder" to make changes

Bonus: We can also leverage this later to introduce the functionality to start over with a site, which a lot of people ask about in support, and we don't have today in WordPress.com. It is explained better here: https://github.com/Automattic/wp-calypso/issues/32792.

What do you think?

olaolusoga commented:
Happy to defer to whatever y'all think is best atm. Don't have bandwidth to dig into this right now. 

I do think if we need to show a modal, there's a flaw somewhere in the UX, and we should improve that vs patching with a modal. We can dig into this with the new onboarding UX.

andrewserong commented:
Thanks @davipontesblog! 

> So to answer those questions we can simplify by offering two options:
> * Start a new site entirely, and scrap the existing one
> * Go back to "Site builder" to make changes

I agree, and what you've described captures the intention I went with in initially going with a modal in this PR, and I like the idea of adding in a "start over" option further down the track. Since it sounds like there isn't scope for design work on this at the moment, does the implicit approach in this PR improve on things, or does it make things harder?

At the very least, with the changes in this PR, hopefully it'll be easier to implement either a modal, or other logic we want to introduce, in the future when we can get it designed.

simison commented:
Can this be closed now that https://github.com/Automattic/wp-calypso/pull/40158 is in?

ramonjd commented:
> Can this be closed now that #40158 is in?

Closing on behalf of @andrewserong (who can reopen if need be)

andrewserong commented:
Thanks for closing, yes let's explore an implementation of what we added in https://github.com/Automattic/wp-calypso/pull/40158 in a separate PR.
